crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -300,12 +355,95 @@ where
 
         Ok(())
     }
+
+    /// Migrates a ModificationHistory key-value pair from V1 to V2.
+    /// The migration fails if other
+    pub fn migrate_modifications_history_at_height(
+        &self,
+        height: u64,
+    ) -> StorageResult<()> {
+        let mut migration_transaction = StorageTransaction::transaction(
+            &self.db,
+            ConflictPolicy::Fail,
+            Changes::default(),
+        );
+
+        let v1_changes = migration_transaction
+            .storage_as_mut::<ModificationsHistoryV1<Description>>()
+            .take(&height)?;
+        if let Some(v1_changes) = v1_changes {
+            migration_transaction
+                .storage_as_mut::<ModificationsHistoryV2<Description>>()
+                .insert(&height, &v1_changes)
+                .unwrap();
+
+            self.db
+                .commit_changes(&migration_transaction.into_changes())?;
+            Ok(())
+        } else {
+            Ok(())
+        }
+    }
+}
+
+// Try to take the value from `ModificationsHistoryV2` first.
+// If the migration is still in progress, remove the value from
+// `ModificationsHistoryV1` and return it if no value for `ModificationsHistoryV2`
+// was found. This is necessary to avoid scenarios where it is possible to
+// roll back twice to the same block height
+fn multiversion_take<Description>(
```

Comment:
`multiversion_take()` and `multiversion_replace()` are very similar, please consider some deduplication. We should be able to provide, let's say, `multiversion_op()` which also takes a generic `F` argument.

Then, on the callsite we can use something like this:
```rust
        let old_changes = multiversion_op(
            storage_transaction,
            height_u64,
            self.migration_in_progress,
            |storage| storage.replace(&height_u64, &reverse_changes),
        )?;
```
or
```rust
        let last_changes = multiversion_op(
            &mut storage_transaction,
            height_to_rollback,
            self.migration_in_progress,
            |storage| storage.take(&height_to_rollback),
        )?
        .ok_or(not_found!(ModificationsHistoryV1<Description>))?;
```

However, due to the complex types, the definition of `F` may become a mess, so I'm leaving it up to you if this is worth changing.
```rust
    F: FnOnce(
        StorageMut<
            '_,
            StructuredStorage<
                InMemoryTransaction<&rocks_db::RocksDb<Historical<Description>>>,
            >,
            ModificationsHistoryV2<Description>,
        >,
    ) -> StorageResult<Option<Changes>>,
```

Reply:
Tried, the signature itself is not a problem, but I spent quite a bit of time fighitng with lifetimes and eventually gave up.

Reply:
Please see here or how I imagined it :)
https://github.com/FuelLabs/fuel-core/pull/2277

But again, up to you. Feel free to close the PR and leave it implemented with two separate functions if you find it more convenient.

Reply:
Thanks, I had a look :) 
I changed this PR quite a lot, so still fighting with typechecking when trying to rebase. 
But I definitely like it how you have done it